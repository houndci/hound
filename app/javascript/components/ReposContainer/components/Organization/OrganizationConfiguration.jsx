import React from "react";
import classNames from "classnames";
import $ from "jquery";

import "object-assign-shim";

import * as Ajax from "../../../../modules/Ajax";

export default class OrganizationConfiguration extends React.Component {
  constructor(props) {
    super(props);

    let repo = "";

    if (props.repo) {
      repo = props.repo;
    }

    this.state = {
      configUpdated: false,
      enabled: !!props.enabled,
      repo
    };
  }

  setEnabled(enabled) {
    this.setState({ enabled });
    this.updateOwner({ config_enabled: enabled });
  }

  setRepo(repo) {
    this.setState({ repo });
    this.updateOwner({ config_repo: repo });
  }

  handleChange(event) {
    this.setRepo(event.target.value);
  }

  toggle() {
    const enabled = !this.state.enabled;
    this.setEnabled(enabled);
  }

  updateOwner(options) {
    const currentState = {
      config_enabled: this.state.enabled,
      config_repo: this.state.repo
    };

    this.setState({ configUpdated: false });
    Ajax.updateOwner(this.props.id, { ...currentState, ...options }).then(() =>
      this.setState({ configUpdated: true })
    );
  }

  renderRepoSelect() {
    if (this.state.enabled) {
      return (
        <div className="organization-header-source">
          <span className="organization-header-label">Use .hound.yml from</span>
          <select
            className="organization-header-select"
            onChange={event => this.handleChange(event)}
            value={this.state.repo}
          >
            {this.renderRepoOptions()}
          </select>
          <div className="inline-flash--success config-enabled">
            {this.state.configUpdated &&
              <span data-role="config-saved">&#10004;</span>}
          </div>
        </div>
      );
    } else {
      return null;
    }
  }

  renderRepoOptions() {
    return this.props.repos.map(repo =>
      <option key={repo.id} value={repo.name}>
        {repo.name}
      </option>
    );
  }

  render() {
    const switchId = `toggle-${this.props.id}`;
    return (
      <div className="organization-header-config">
        <div className="organization-header-toggle">
          <span className="organization-header-label">
            Use organization-wide config
          </span>

          <input
            id={switchId}
            checked={this.state.enabled}
            className="organization-header-toggle-input"
            onChange={() => this.toggle()}
            type="checkbox"
            name="toggle"
          />

          <label className="toggle-switch" htmlFor={switchId} />
        </div>

        {this.renderRepoSelect()}
      </div>
    );
  }
}
