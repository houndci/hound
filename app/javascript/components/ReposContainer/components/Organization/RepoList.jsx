import React from 'react';

import EmptyRepoList from './RepoList/EmptyRepoList';
import PopulatedRepoList from './RepoList/PopulatedRepoList';

export default class RepoList extends React.Component {
  render() {
    const {
      repos,
      onRepoClicked,
      isProcessingId,
      filterTerm,
    } = this.props;

    if (repos.length > 0) {
      return (
        <PopulatedRepoList
          repos={repos}
          onRepoClicked={onRepoClicked}
          isProcessingId={isProcessingId}
          filterTerm={filterTerm}
        />
      );
    } else {
      return (
        <EmptyRepoList />
      );
    }
  }
}
