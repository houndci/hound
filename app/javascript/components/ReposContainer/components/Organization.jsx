import React from 'react';

import OrganizationConfiguration from "./Organization/OrganizationConfiguration";
import RepoList from './Organization/RepoList';

export default class Organization extends React.Component {
  render() {
    const {
      name,
      onRepoClicked,
      isProcessingId,
      repos,
      configEnabled,
      configRepo,
      filterTerm,
      ownerId,
    } = this.props;

    return (
      <div className="organization" data-org-name={name}>
        <header className="organization-header">
          <h2 className="organization-header-title">{name}</h2>

          <OrganizationConfiguration
            enabled={configEnabled}
            id={ownerId}
            repo={configRepo}
            repos={repos}
          />
        </header>
        <section className="repo_listing">
          <RepoList
            repos={repos}
            onRepoClicked={onRepoClicked}
            isProcessingId={isProcessingId}
            filterTerm={filterTerm}
          />
        </section>
      </div>
    );
  }
}
