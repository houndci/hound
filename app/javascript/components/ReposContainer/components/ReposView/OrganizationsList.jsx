import React from 'react';
import _ from 'lodash';

import Organization from '../Organization';

export default class OrganizationsList extends React.Component {
  reposForOrg(org) {
    if (_.has(org, "id")) {
      return _.filter(this.props.repos, (repo) => {
        return repo.owner.id === org.id;
      });
    } else {
      return _.filter(this.props.repos, (repo) => {
        return this.orgName(repo.name) === org.name;
      });
    }
  }

  orgName(name) {
    return _.split(name, "/")[0];
  }

  render() {
    const {
      repos,
      onRepoClicked,
      filterTerm,
      isProcessingId,
      organizations,
    } = this.props;

    return (
      <section className="organizations">
        {organizations.map( org => (
          <Organization
            name={org.name}
            key={org.id || org.name}
            repos={this.reposForOrg(org)}
            onRepoClicked={onRepoClicked}
            filterTerm={filterTerm}
            isProcessingId={isProcessingId}
            configEnabled={org.config_enabled}
            configRepo={org.config_repo}
            ownerId={org.id}
          />
        ))}
      </section>
    );
  }
}
