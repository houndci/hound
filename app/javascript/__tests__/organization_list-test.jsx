import OrganizationsList from "../components/ReposContainer/components/ReposView/OrganizationsList";
import Organization from "../components/ReposContainer/components/Organization";

it("renders a list of organizations appropriately", () => {
  const organizations = [
    { id: 1, name: "Test org" }
  ];
  const repos = [
    {
      id: 1,
      name: "Test repo",
      owner: {
        id: 1
      }
    }
  ];

  const wrapper = shallow(
    <OrganizationsList
      organizations={organizations}
      repos={repos}
      filterTerm={""}
      onRepoClicked={jest.fn()}
      isProcessingId={null}
    />
  );
  expect(wrapper).toMatchSnapshot();
});

it("filters a list of organizations appropriately", () => {
  const organizations = [
    { id: 1, name: "Test org 1" },
    { id: 2, name: "Test org 2" },
    { id: 3, name: "Test org 3" },
  ];

  const repos = [
    {
      id: 1,
      name: "Test repo",
      owner: {
        id: 2
      }
    }
  ];

  const wrapper = mount(
    <OrganizationsList
      organizations={organizations}
      repos={repos}
      filterTerm={""}
      onRepoClicked={jest.fn()}
      isProcessingId={null}
    />
  );
  const organizationsFound = wrapper.find(Organization).nodes;
  expect(organizationsFound.length).toBe(3);

  expect(organizationsFound[0].props.name).toBe("Test org 1");
  expect(organizationsFound[0].props.repos.length).toBe(0);

  expect(organizationsFound[1].props.name).toBe("Test org 2");
  expect(organizationsFound[1].props.repos.length).toBe(1);
  expect(organizationsFound[1].props.repos[0].id).toBe(1);

  expect(organizationsFound[2].props.name).toBe("Test org 3");
  expect(organizationsFound[2].props.repos.length).toBe(0);
});

it("filters a list of organizations appropriately", () => {
  const organizations = [
    { name: "S.H.I.E.L.D." },
    { name: "H.A.M.M.E.R." },
    { name: "Hydra" },
  ];

  const repos = [
    {
      id: 666,
      name: "Hydra/deathray",
    }
  ];

  const wrapper = mount(
    <OrganizationsList
      organizations={organizations}
      repos={repos}
      filterTerm={""}
      onRepoClicked={jest.fn()}
      isProcessingId={null}
    />
  );
  const organizationsFound = wrapper.find(Organization).nodes;
  expect(organizationsFound.length).toBe(3);

  expect(organizationsFound[0].props.name).toBe("S.H.I.E.L.D.");
  expect(organizationsFound[0].props.repos.length).toBe(0);

  expect(organizationsFound[1].props.name).toBe("H.A.M.M.E.R.");
  expect(organizationsFound[1].props.repos.length).toBe(0);

  expect(organizationsFound[2].props.name).toBe("Hydra");
  expect(organizationsFound[2].props.repos.length).toBe(1);
  expect(organizationsFound[2].props.repos[0].name).toBe("Hydra/deathray");
});
