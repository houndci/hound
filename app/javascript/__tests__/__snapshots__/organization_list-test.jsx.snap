exports[`test renders a list of organizations appropriately 1`] = `
<section
  className="organizations">
  <Organization
    filterTerm=""
    isProcessingId={null}
    name="Test org"
    onRepoClicked={[Function]}
    ownerId={1}
    repos={
      Array [
        Object {
          "id": 1,
          "name": "Test repo",
          "owner": Object {
            "id": 1,
          },
        },
      ]
    } />
</section>
`;
